plugins/inputs/t128_graphql/t128_graphql.go

```diff
@@ -0,0 +1,284 @@
+
+	for _, element := range plugin.Fields {
+		query += "\n" + element
+	}
+	query = strings.TrimSpace(query)
+	for _, element := range plugin.Tags {
+		query += "\n" + element
+	}
+
+	query += "\n" + strings.Repeat("}", strings.Count(query, "{"))
+
+	return query
+}
+
+func (plugin *T128GraphQL) retrieveMetrics(acc telegraf.Accumulator, timestamp time.Time) {
+	request, err := plugin.createRequest()
+	if err != nil {
+		acc.AddError(fmt.Errorf("failed to create a request for query %s: %w", plugin.Query, err))
+		return
+	}
+
+	response, err := plugin.Client.Do(request)
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(response.Body)
+
+	if err != nil {
+		template := "query error for metric %s: %s"
+		plugin.DecodeAndReportJsonErrors(buf.Bytes(), acc, template)
+		return
+	}
+	defer response.Body.Close()
+
+	if response.StatusCode < 200 || response.StatusCode >= 300 {
+		template := fmt.Sprintf("status code %d not OK for metric ", response.StatusCode) + "%s: %s"
+		plugin.DecodeAndReportJsonErrors(buf.Bytes(), acc, template)
+		return
+	}
+
+	//decode json
+	jsonParsed, err := gabs.ParseJSON(buf.Bytes())
+	if err != nil {
+		acc.AddError(fmt.Errorf("invalid json response for metric %s: %w", plugin.CollectorName, err))
+		return
+	}
+
+	jsonEntryPoint := plugin.BuildJsonEntryPoint()
+	jsonObj, err := jsonParsed.JSONPointer(jsonEntryPoint)
+	if err != nil {
+		template := "unexpected response for metric %s: %s"
+		plugin.DecodeAndReportJsonErrors(buf.Bytes(), acc, template)
+		return
+	}
+
+	//update acc
+	jsonChildren, err := jsonObj.Children()
+	if err != nil {
+		acc.AddError(fmt.Errorf("failed to iterate on response nodes for metric %s: %w", plugin.CollectorName, err))
+		return
+	}
+
+	//TODO: allow nested fields/tags
+	for _, child := range jsonChildren {
+		node := child.Data().(map[string]interface{})
+		fields := make(map[string]interface{})
+		tags := make(map[string]string)
+
+		//allow integer conversion?
+		//TODO: fully implement multiple fields
+		for fieldRenamed, fieldName := range plugin.Fields {
+			if isNil(node[fieldName]) {
+				acc.AddError(fmt.Errorf("found empty data for metric %s: field %s", plugin.CollectorName, fieldName))
+				continue
+			}
+			fields[fieldRenamed] = node[fieldName]
+
+			for tagRenamed, tagName := range plugin.Tags {
+				if isNil(node[tagName]) {
+					tags[tagRenamed] = tagRenamed
+					continue
+				}
+				//gabs module converts data to either string or float64
+				strTag, ok := node[tagName].(string)
+				if !ok {
+					floatTag := node[tagName].(float64)
+					strTag = strconv.FormatFloat(floatTag, 'f', -1, 64)
+				}
+				tags[tagRenamed] = strTag
+			}
+
+			acc.AddFields(
+				plugin.CollectorName,
+				fields,
+				tags,
+				timestamp,
+			)
+		}
+	}
+}
+
+func (plugin *T128GraphQL) createRequest() (*http.Request, error) {
+	content := struct {
+		Query string `json:"query,omitempty"`
+	}{
+		plugin.Query,
+	}
+
+	//inject env config?
```

Comment:
is this supposed to be a TODO? What do other input collectors do?

Reply:
I think other collectors inject env config at the monitoring-agent level so I left this here to make sure with you all

Reply:
What env config?

Reply:
In `arp_state.py` I see this `data = graphql.execute_query(self.base_url, self.query, variables=self.env_config)`. I meant those `env_config`

Reply:
Gotcha - that might be something else. I guess a good question would be whether we want the monitoring agent to do the substitution based on the environment or we still leverage the environment concept in telegraf. This goes to my other questions - do we know if any of the other telegraf inputs do something with the `telegraf` environmnent. @gregschrock did we use it in the metric collector for example?